lang/java/avro/src/main/java/org/apache/avro/file/DataFileReader.java

```diff
@@ -312,10 +312,10 @@ public long skip(long skip) throws IOException {
       long length = in.length();
       long remaining = length - position;
       if (remaining > skip) {
-        in.seek(skip);
+        in.seek(position + skip);
         return in.tell() - position;
```

Comment:
And in same time, either `return skip; ... else { ... return length - position; }`  either put `return in.tell() - position;` at the end of function.